clients/drcachesim/tools/filter/type_filter.h

```diff
@@ -0,0 +1,110 @@
+ *   this list of conditions and the following disclaimer.
+ *
+ * * Redistributions in binary form must reproduce the above copyright notice,
+ *   this list of conditions and the following disclaimer in the documentation
+ *   and/or other materials provided with the distribution.
+ *
+ * * Neither the name of Google, Inc. nor the names of its contributors may be
+ *   used to endorse or promote products derived from this software without
+ *   specific prior written permission.
+ *
+ * THIS SOFTWARE IS PROVIDED BY THE COPYRIGHT HOLDERS AND CONTRIBUTORS "AS IS"
+ * AND ANY EXPRESS OR IMPLIED WARRANTIES, INCLUDING, BUT NOT LIMITED TO, THE
+ * IMPLIED WARRANTIES OF MERCHANTABILITY AND FITNESS FOR A PARTICULAR PURPOSE
+ * ARE DISCLAIMED. IN NO EVENT SHALL VMWARE, INC. OR CONTRIBUTORS BE LIABLE
+ * FOR ANY DIRECT, INDIRECT, INCIDENTAL, SPECIAL, EXEMPLARY, OR CONSEQUENTIAL
+ * DAMAGES (INCLUDING, BUT NOT LIMITED TO, PROCUREMENT OF SUBSTITUTE GOODS OR
+ * SERVICES; LOSS OF USE, DATA, OR PROFITS; OR BUSINESS INTERRUPTION) HOWEVER
+ * CAUSED AND ON ANY THEORY OF LIABILITY, WHETHER IN CONTRACT, STRICT
+ * LIABILITY, OR TORT (INCLUDING NEGLIGENCE OR OTHERWISE) ARISING IN ANY WAY
+ * OUT OF THE USE OF THIS SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF SUCH
+ * DAMAGE.
+ */
+
+#ifndef _TYPE_FILTER_H_
+#define _TYPE_FILTER_H_ 1
+
+#include "record_filter.h"
+#include "trace_entry.h"
+
+#include <vector>
+#include <unordered_set>
+
+#ifdef MACOS
+// Provide trace_type_t and trace_marker_type_t specializations to
+// allow declaration of unordered_set<_> below. This was needed for
+// OSX particularly. In C++14, std::hash works as expected for enums
+// too: https://cplusplus.com/forum/general/238538/.
+namespace std {
+template <> struct hash<trace_type_t> {
+    size_t
+    operator()(trace_type_t t) const
+    {
+        return static_cast<size_t>(t);
+    }
+};
+template <> struct hash<trace_marker_type_t> {
+    size_t
+    operator()(trace_marker_type_t t) const
+    {
+        return static_cast<size_t>(t);
+    }
+};
+} // namespace std
+#endif
+
+namespace dynamorio {
+namespace drmemtrace {
+
+class type_filter_t : public record_filter_t::record_filter_func_t {
+public:
+    type_filter_t(std::vector<trace_type_t> remove_trace_types,
+                  std::vector<trace_marker_type_t> remove_marker_types)
+    {
+        for (auto trace_type : remove_trace_types) {
+            remove_trace_types_.insert(trace_type);
+        }
+        for (auto marker_type : remove_marker_types) {
+            remove_marker_types_.insert(marker_type);
+        }
+    }
+    void *
+    parallel_shard_init(memtrace_stream_t *shard_stream) override
+    {
+        return nullptr;
+    }
+    bool
+    parallel_shard_filter(const trace_entry_t &entry, void *shard_data) override
+    {
+        if (remove_trace_types_.find(static_cast<trace_type_t>(entry.type)) !=
+            remove_trace_types_.end()) {
+            if (entry.type != TRACE_TYPE_MARKER)
```

Comment:
I don't understand this: the user is supposed to pass the marker type to the list to be removed even if not all markers will be removed and even though there is a dedicated list of marker sub-types to remove?  It seems more intuitive and simpler to *not* pass the marker type and *not* need this special check here for that use case.  It needs to be documented regardless since as a user I would not expect this behavior.  If I want to remove *all* marker types, I would want to be able to do that by passing the marker type to the remove_trace_types list and *not* have to list every single possible type (thus making my tool not forward-compatible).

Reply:
Makes sense. Changed it.